lib/neptune/federated_learning/aggregator.py

```diff
@@ -0,0 +1,209 @@
+            await stop
+            LOG.info("websocket server close")
+
+    async def _receive(self, websocket, path):
+        """Receive the data form worker and aggregate, then send the result
+            back to to worker
+        """
+        async for json_data in websocket:
+            agg = AggregationData.from_json(json_data)
+            self.aggregator.agg_data_dict[agg.worker_id] = agg
+            self.ws_clients[agg.worker_id] = websocket
+            if self.aggregator.current_worker is None:
+                self.aggregator.current_worker = agg.worker_id
+                LOG.info(f"current aggregation worker_id is {agg.worker_id}")
+            LOG.info(f"received agg data, "
+                     f"task_name={agg.task_id}, "
+                     f"worker_id={agg.worker_id}, "
+                     f"agg_data_dict length "
+                     f"is {len(self.aggregator.agg_data_dict)}")
+
+            agg_data_dict_len = len(self.aggregator.agg_data_dict)
+            if agg_data_dict_len == self.config.participants_count:
+                self.aggregator.aggregate()
+                self.aggregator.update_task_info()
+                self.aggregator.exit_check()
+                items = self.aggregator.agg_data_dict_aggregated.items()
+                # use coroutine the send the data
+                tasks = []
+                for worker_id, agg_data in items:
+                    tasks.append(asyncio.ensure_future(
+                        self.ws_clients[worker_id].send(agg_data.to_json())
+                    ))
+                    LOG.info("send agg_data to worker, "
+                             f"worker_id = {worker_id}")
+
+                # wait for all task complete
+                for task in tasks:
+                    await task
+
+                # record the next round start time, after all agg data
+                # transmitted.
+                self.aggregator.start_time = time.strftime("%Y-%m-%d %H:%M:%S",
+                                                           time.localtime())
+
+                if self.aggregator.exit_flag:
+                    LOG.info(f"aggregation finished")
+
+
+class AggregationClient:
+    """Client that interacts with the cloud aggregator."""
+    _retry = 15
+    _retry_interval_seconds = 3
+    _ws_timeout = 5
+
+    def __init__(self, ip, port):
+        self.ip = ip
+        self.port = port
+        self.agg_data = None
+        self.ws_url = f"ws://{self.ip}:{self.port}"
+
+    async def _update_weights(self, agg_data):
+        error = None
+        for i in range(AggregationClient._retry):
+            try:
+                # TODO should not connect websocket each time, use long
+                # connection
+                websocket = await asyncio.wait_for(websockets.connect(
+                    self.ws_url,
+                    max_size=MAX_SIZE_BYTE
+                ), self._ws_timeout)
+                await websocket.send(agg_data.to_json())
```

Comment:
agg_data.to_json() should be cached in case too large.

Reply:
why to_json in that case websocket.send needs to serialize again?  why not just to_bytes?